frontend/src/test/java/com/google/sps/InMemoryPlaceVisitStorageTest.java

```diff
@@ -0,0 +1,333 @@
+    .setLatitude(35.32)
+    .setLongitude(125.33)
+    .build();
+
+  private static final PlaceVisitModel SEOUL = PlaceVisitModel.builder()
+    .setPlaceId("ICN")
+    .setName("Seoul, South Korea")
+    .setTripId("b")
+    .setUserMark("if-time")
+    .setLatitude(26.32)
+    .setLongitude(135.33)
+    .build();
+
+  private static final PlaceVisitModel TOKYO_B = PlaceVisitModel.builder()
+    .setPlaceId("HND")
+    .setName("Tokyo, Japan")
+    .setTripId("b")
+    .setUserMark("must-see")
+    .setLatitude(35.32)
+    .setLongitude(139.33)
+    .build();
+
+  private static final PlaceVisitModel SEOUL_B = PlaceVisitModel.builder()
+    .setPlaceId("ICN")
+    .setName("some city in Korea")
+    .setTripId("b")
+    .setUserMark("if-time")
+    .setLatitude(0)
+    .setLongitude(0)
+    .build();
+
+  /** 
+   * test that a PlaceVisitAlreadyExistsException is thrown when a place that 
+   * has already been added to the storage is added again 
+   */
+  @Test(expected = PlaceVisitAlreadyExistsException.class)
+  public void testAddSamePlaceVisit_throwsException() throws PlaceVisitAlreadyExistsException {
+    InMemoryPlaceVisitStorage storage = new InMemoryPlaceVisitStorage();
+    storage.addPlaceVisit(LONDON);
+    storage.addPlaceVisit(LONDON);
+  }
+
+  /** 
+   * test that a PlaceVisitAlreadyExistsException is thrown when a place is  
+   * added but another place with the same placeId and tripId has already been 
+   * added to the storage
+   */
+  @Test(expected = PlaceVisitAlreadyExistsException.class)
+  public void testAddPlaceVisit_samePlaceIdAndTripId_throwsException() throws PlaceVisitAlreadyExistsException {
+    InMemoryPlaceVisitStorage storage = new InMemoryPlaceVisitStorage();
+    storage.addPlaceVisit(SEOUL);
+    storage.addPlaceVisit(SEOUL_B);
+  }
+
+  /**
+   * test that an exception is not thrown when two PlaceVisitModel objects
+   * with different tripId and/or placeId are added to storage
+   */
+  @Test
+  public void testAddSamePlaceDiffTrip() throws PlaceVisitAlreadyExistsException {
+    InMemoryPlaceVisitStorage storage = new InMemoryPlaceVisitStorage();
+    storage.addPlaceVisit(TOKYO);
+    storage.addPlaceVisit(TOKYO_B);
+  }
+
+  /**
+   * test that a PlaceNotFoundException is thrown when getting a place from
+   * an empty storage
+   */
+  @Test(expected = PlaceVisitNotFoundException.class)
+  public void testGetPlace_emptyStorage() throws PlaceVisitNotFoundException {
+    NO_PLACE_VISITS.getPlaceVisit(TOKYO.tripId(), TOKYO.placeId());
+  }
+
+  /**
+   * test that a PlaceNotFoundException is thrown when trying to get a place
+   * from storage when there is no place in storage with the given tripId and
+   * placeId
+   */
+  @Test(expected = PlaceVisitNotFoundException.class)
+  public void testGetPlace_notInStorage() throws PlaceVisitAlreadyExistsException, PlaceVisitNotFoundException {
+    InMemoryPlaceVisitStorage storage = new InMemoryPlaceVisitStorage();
+    storage.addPlaceVisit(LONDON);
```

Comment:
nit: not sure why all these adds are necessary. I can see adding one to test the difference between the empty and the "something stored but not with this ID" cases, but this seems like overkill.